Implement basic HTTP authentication support

tadeboro opened the pull request:
Up until now, Redfish client only supported session authentication,
since it is mandatory as per Redfish standard [1]. This means that
client can interact with any Redfish-compliant service.

Unfortunately, services in the wild often only implement part of the
standard and in case of authentication, some of the services only
offer basic authentication, which makes our client useless for such
cases.

To bypass this, we added basic authentication support. By default, we
try to use session authentication if possible, since it is more secure
of the two, and only use basic authentication if session cannot be
created.

[1] http://redfish.dmtf.org/schemas/DSP0266_1.4.0.html#authentication

gberginc commented on spec/redfish_client/root_spec.rb:
```diff
@@ -46,24 +58,49 @@
-      expect(subject.Auth.key).to eq("val")
+    context "#login" do
+      it "authenticates user against service" do
+        subject.login("user", "pass")
```
Would you consider using `before` block within `context` which would contain the call to login?

```
before do
    subject.login('user', 'pass')
end
```

gberginc commented on spec/redfish_client/root_spec.rb:
```diff
@@ -46,24 +58,49 @@
     end
   end
 
+  let(:connector) { RedfishClient::Connector.new("http://example.com") }
```
Is this something that needs to be on a "global" level? Can you create within `context` like in L62?

tadeboro replied:
Connectors are passive objects that can be freely reused, so having them defined here should do no harm. Exceptions to this rules are login and logout methods that do modify the connector by adding/removing some headers and this is why tests for those methods receive dedicated one in order not to interfere with the rest of the tests.

Thinking about this, it would be even cleaner to create new connector for each new created resource. That way test would share no state at all. I will create new pull request that will contain this cleanup throughout the tests.

gberginc commented on lib/redfish_client/root.rb:
```diff
@@ -51,5 +49,37 @@ def logout
+    end
+
+    def basic_login(username, password)
+      payload = Base64.encode64("#{username}:#{password}").strip
```
Would it make sense to actually do a request afterwards to try this basic auth? In session login, we will get a report on whether the username:pass is incorrect, however with Basic we have no such status report following the login.

tadeboro replied:
I had a call to `self.Systems` here, but then I found a Redfish service that does not expose this endpoint (although Redfish standard mandates it). So selecting an endpoint to query after the authentication is not completely trivial.

Two options to securely implement login verification I see here are:

 1. fix an endpoint that should exist in order for service to be usable (like aforementioned `Systems`) or
 2. iterate root resource, find first usable `@odata.id` field, query it and check for status 200.

Second option will allow us to use just about any Redfish service in exchange for some complications is client code. Is it worth it or should we go with first option?

gberginc commented on lib/redfish_client/root.rb:
```diff
@@ -9,39 +10,36 @@ class Root < Resource
+      # only if session auth is not available.
+      begin
+        session_login(username, password)
+      rescue
```
Question: in case I choose wrong user:pass pair, `session_login` will raise an error and will attempt the basic login. Current logic does not check whether basic login actually works (see also my comment below). This means that after invalid user:pass pair, the client will be configure to use Basic auth method, but will fail for session-based services on the first call...

Thus:
 * is there a way to know whether the session-based auth failed because of it not being supported or because the user:pass pair is illegal?
 * can you handle the basic auth and check that it actually works with the given user:pass pair?

tadeboro replied:
We can easily determine if session auth is supported or not (we just need to check for `SessonService` singleton presence), so differentiating between invalid password and not being able to create session is trivial.

Basic authentication is another beast altogether. In theory, we can determine if the service supports basic authentication by looking for `WWW-Authenticate` HTTP header, but there is no guarantee that service will return it (again, this is mandated by Redfish standard, but reality tells a different story). In any case, we can tell user that username and/or password is invalid for services that support authentication, but authenticationg against the service that has no authentication will always succeed.

tadeboro commented:
I addressed all of the comments with the exception of the global `connector` test variable that I would prefer to fix in another pull request as suggested in my inline response.

@gberginc Does this sound sensible?
